core_new/src/message/context.rs

```diff
@@ -0,0 +1,313 @@
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::identifiers::{MessageId, ProgramId};
+use crate::message::{
+    Dispatch, HandleMessage, HandlePacket, IncomingMessage, InitMessage, InitPacket, Payload,
+    ReplyMessage, ReplyPacket,
+};
+use alloc::{
+    collections::{BTreeMap, BTreeSet},
+    vec::Vec,
+};
+use codec::{Decode, Encode};
+use scale_info::TypeInfo;
+
+/// Context settings.
+#[derive(Copy, Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd, Decode, Encode, TypeInfo)]
+pub struct ContextSettings {
+    sending_fee: u64,
+    outgoing_limit: u32,
+}
+
+impl ContextSettings {
+    /// Create new ContextSettings.
+    pub fn new(sending_fee: u64, outgoing_limit: u32) -> Self {
+        Self {
+            sending_fee,
+            outgoing_limit,
+        }
+    }
+}
+
+impl Default for ContextSettings {
+    fn default() -> Self {
+        Self::new(0, 1024)
+    }
+}
+
+/// Context outcome.
+///
+/// Contains all sendings and wakes that should be done after execution.
+#[derive(Default, Debug, Eq, Hash, Ord, PartialEq, PartialOrd, Decode, Encode, TypeInfo)]
+pub struct ContextOutcome {
+    init: Vec<InitMessage>,
+    handle: Vec<HandleMessage>,
+    reply: Option<ReplyMessage>,
+    awakening: Vec<MessageId>,
+    // Additional information section
+    program_id: ProgramId,
+    source: ProgramId,
+    origin_msg_id: MessageId,
+}
+
+impl ContextOutcome {
+    /// Create new ContextOutcome.
+    fn new(program_id: ProgramId, source: ProgramId, origin_msg_id: MessageId) -> Self {
+        Self {
+            program_id,
+            source,
+            origin_msg_id,
+            ..Default::default()
+        }
+    }
+
+    /// Destructs outcome after execution and returns provided dispatches and awaken message ids.
+    pub fn drain(self) -> (Vec<Dispatch>, Vec<MessageId>) {
+        let mut dispatches = Vec::new();
+
+        if let Some(msg) = self.reply {
+            dispatches.push(msg.into_dispatch(self.program_id, self.source, self.origin_msg_id));
+        };
+
+        for msg in self.init.into_iter() {
+            dispatches.push(msg.into_dispatch(self.program_id));
+        }
+
+        for msg in self.handle.into_iter() {
+            dispatches.push(msg.into_dispatch(self.program_id));
+        }
+
+        (dispatches, self.awakening)
+    }
+}
+
+/// Store of previous message execution context.
+#[derive(Clone, Default, Debug, Eq, Hash, Ord, PartialEq, PartialOrd, Decode, Encode, TypeInfo)]
+pub struct ContextStore {
+    outgoing: BTreeMap<u32, Option<Payload>>,
+    reply: Option<Payload>,
+    initialized: BTreeSet<ProgramId>,
```

Comment:
`Vec` was used intentionally here, because during one execution we can have multiple program creation calls in some order, that should be kept in `processor`, dispatch queue and other layers.

Reply:
In order to make search for duplicates more efficient we can find some _O(n*logn)_, or even _O(n)_solutions, if possible.

Reply:
Because `initialized` is used only for checks, this is resolved.